src/Hosting/TestHost/src/ClientHandler.cs

```diff
@@ -115,28 +148,39 @@ internal ClientHandler(PathString pathBase, ApplicationWrapper application)
                     }
                 }
 
-                if (body.CanSeek)
-                {
-                    // This body may have been consumed before, rewind it.
-                    body.Seek(0, SeekOrigin.Begin);
-                }
-                req.Body = new AsyncStreamWrapper(body, () => contextBuilder.AllowSynchronousIO);
+                req.Body = new AsyncStreamWrapper(requestPipe.Reader.AsStream(), () => contextBuilder.AllowSynchronousIO);
             });
 
             var response = new HttpResponseMessage();
 
             // Copy trailers to the response message when the response stream is complete
-            contextBuilder.RegisterResponseReadCompleteCallback(context =>
+            contextBuilder.RegisterResponseReadCompleteCallback(async context =>
             {
-                var responseTrailersFeature = context.Features.Get<IHttpResponseTrailersFeature>();
+                try
+                {
+                    var responseTrailersFeature = context.Features.Get<IHttpResponseTrailersFeature>();
+
+                    foreach (var trailer in responseTrailersFeature.Trailers)
+                    {
+                        bool success = response.TrailingHeaders.TryAddWithoutValidation(trailer.Key, (IEnumerable<string>)trailer.Value);
+                        Contract.Assert(success, "Bad trailer");
+                    }
 
-                foreach (var trailer in responseTrailersFeature.Trailers)
+                    await copyRequestTask;
+                    await requestPipe.Reader.CompleteAsync();
```

Comment:
Why does the response body block completion and wait for the request body to drain? What if the server didn't want to read the request body? See my server side abort question.

The app could always drain the request body if it wanted to, but it's not mandatory. The copyRequestTask would also block on buffer space if the app isn't reading it, so this could hang.

Reply:
We need to await the task somewhere, there shouldn't be anything doing a fire and forget.

Reply:
Yes, but we also need to guarantee that it completes or gets cancelled. Doing this at the end of the response body doesn't make sense regardless. When the app finishes would be better: https://github.com/aspnet/AspNetCore/blob/0da5fabdbdd462cdd9528398efb6b1ffb9b27f65/src/Hosting/TestHost/src/HttpContextBuilder.cs#L98-L105 is probob

Reply:
Agreed. Actually I initially told @JamesNK to put it there 😄

Reply:
Done